dll/win32/shell32/wine/pidl.c

```diff
@@ -1722,6 +1722,62 @@ LPITEMIDLIST _ILCreateGuidFromStrA(LPCSTR szGUID)
 }
 #endif
 
+#ifdef __REACTOS__
+static const struct {
+    LPCWSTR pszName;
+    LPCGUID guid;
+} ShellShortcuts[] =
+{
+    {L"ADMINISTRATIVE TOOLS", &CLSID_AdminFolderShortcut},
+    {L"CONTROLPANELFOLDER", &CLSID_ControlPanel},
+    {L"FONTS", &CLSID_FontsFolderShortcut},
+    {L"MYCOMPUTERFOLDER", &CLSID_MyComputer},
+    {L"PERSONAL", &CLSID_MyDocuments},
+    {L"NETWORKPLACESFOLDER", &CLSID_NetworkPlaces},
+    {L"CONNECTIONSFOLDER", &CLSID_ConnectionFolder},
+    {L"PRINTERSFOLDER", &CLSID_Printers},
+    {L"RECYCLEBINFOLDER", &CLSID_RecycleBin}
+};
+
+// Create GUID from shortcut(desktop, Personal...)
+// Returns TRUE if successful and FALSE otherwise.
+static BOOL _ILCreateGuidFromShortcutW(LPCWSTR szShortcut, GUID *fidOut)
+{
+    UINT i;
+    for (i = 0; i < _countof(ShellShortcuts); ++i)
+    {
+        if (StrCmpIW(szShortcut, ShellShortcuts[i].pszName) == 0)
+        {
+            *fidOut = *ShellShortcuts[i].guid;
+            return TRUE;
+        }
+    }
+    return FALSE;
+}
+
+// If szUrl is '::{GUID}' then returns a GUID
+// otherwise it tries to resolve the shortcut.
+LPITEMIDLIST _ILCreateGuidFromShellUrlW(LPCWSTR szUrl)
+{
+    GUID iid;
+
+    if (szUrl[0] == L':' && szUrl[1] == L':')
```

Comment:
What if only 1 character? Eg [1] does not exists ?

Reply:
what is 1 character? it checks [0] and [1]

Reply:
What KRosUser2 means, is what happens if for some reason the helper function is called with a one-character NULL-terminated (or just an empty NULL-terminated) string. In that very latter case, only szUrl[0] would be valid (and be the NULL char) while szUrl[1] would be undefined and you could crash the code.

Reply:
Their initial point probably was that if the string is empty the szUrl[0] == L':' check will fail and the szUrl[1] == L':' check isn't going to happen